packager/util/config_test.go

```diff
@@ -0,0 +1,267 @@
+		CertFile = "cert.pem"
+		KeyFile = "key.pem"
+		OCSPCache = "/tmp/ocsp"
+		[[URLSet]]
+		  [URLSet.Sign]
+		    Domain = "example.com"
+		    PathExcludeRE = ["["]
+	`))), "PathExcludeRE contains invalid regexp")
+}
+
+func TestInvalidQueryRE(t *testing.T) {
+	assert.Contains(t, errorFrom(ReadConfig([]byte(`
+		CertFile = "cert.pem"
+		KeyFile = "key.pem"
+		OCSPCache = "/tmp/ocsp"
+		[[URLSet]]
+		  [URLSet.Sign]
+		    Domain = "example.com"
+		    QueryRE = "["
+	`))), "QueryRE must be a valid regexp")
+}
+
+func TestSignMissing(t *testing.T) {
+	msg := errorFrom(ReadConfig([]byte(`
+		CertFile = "cert.pem"
+		KeyFile = "key.pem"
+		OCSPCache = "/tmp/ocsp"
+		[[URLSet]]
+	`)))
+	assert.Contains(t, msg, "This section must be specified")
+	assert.Contains(t, msg, "URLSet.0.Sign")
+}
+
+func TestSignScheme(t *testing.T) {
+	assert.Contains(t, errorFrom(ReadConfig([]byte(`
+		CertFile = "cert.pem"
+		KeyFile = "key.pem"
+		OCSPCache = "/tmp/ocsp"
+		[[URLSet]]
+		  [URLSet.Sign]
+		    Domain = "example.com"
+		    Scheme = ["http"]
+	`))), "Scheme not allowed here")
+}
+
+func TestSignDomainRE(t *testing.T) {
+	assert.Contains(t, errorFrom(ReadConfig([]byte(`
+		CertFile = "cert.pem"
+		KeyFile = "key.pem"
+		OCSPCache = "/tmp/ocsp"
+		[[URLSet]]
+		  [URLSet.Sign]
+		    Domain = "example.com"
+		    DomainRE = ".*"
+	`))), "DomainRE not allowed here")
+}
+
+func TestSignSamePath(t *testing.T) {
+	assert.Contains(t, errorFrom(ReadConfig([]byte(`
+		CertFile = "cert.pem"
+		KeyFile = "key.pem"
+		OCSPCache = "/tmp/ocsp"
+		[[URLSet]]
+		  [URLSet.Sign]
+		    Domain = "example.com"
+		    SamePath = true
+	`))), "SamePath not allowed here")
+}
+
+func TestSignOverrides(t *testing.T) {
+	config, err := ReadConfig([]byte(`
+		CertFile = "cert.pem"
+		KeyFile = "key.pem"
+		OCSPCache = "/tmp/ocsp"
+		[[URLSet]]
+		  [URLSet.Sign]
+		    Domain = "example.com"
+		    PathRE = "/amp/.*"
+		    PathExcludeRE = ["/amp/signin", "/amp/settings(/.*)?"]
+		    QueryRE = ""
+		    ErrorOnStatefulHeaders = true
+	`))
+	require.NoError(t, err)
+	require.Equal(t, 1, len(config.URLSet))
+	// TODO(twifkak): Don't depend on scheme order.
+	assert.Equal(t, URLPattern{
+		Domain: "example.com",
+		PathRE: stringPtr("/amp/.*"),
+		PathExcludeRE: []string{"/amp/signin", "/amp/settings(/.*)?"},
+		QueryRE: stringPtr(""),
+		ErrorOnStatefulHeaders: true,
+	}, *config.URLSet[0].Sign)
+}
+
+func TestFetchDefaults(t *testing.T) {
+	config, err := ReadConfig([]byte(`
+		CertFile = "cert.pem"
+		KeyFile = "key.pem"
+		OCSPCache = "/tmp/ocsp"
+		[[URLSet]]
+		  [URLSet.Sign]
+		    Domain = "example.com"
+		  [URLSet.Fetch]
+		    Domain = "example.com"
+	`))
+	require.NoError(t, err)
+	require.Equal(t, 1, len(config.URLSet))
+	fetch := *config.URLSet[0].Fetch
+	assert.ElementsMatch(t, []string{"http", "https"}, fetch.Scheme)
+	assert.Equal(t, "", fetch.DomainRE)
+	assert.Equal(t, "example.com", fetch.Domain)
+	assert.Equal(t, stringPtr(".*"), fetch.PathRE)
```

Comment:
Yep, it passes. assert.Equal does deep equality (see https://godoc.org/github.com/stretchr/testify/assert#Equal). This even goes for pointers inside of structs.

Reply:
For simple primitives, this seems overkill (reflect deep equals plus the func definitions for string/boolPtr).

Reply:
The downside of `assert.Equal(t. ".*", *fetch.PathRE)` is that if `fetch.PathRE` is nil, I get a panic instead of a graceful test failure. I could do:
```
require.NotNil(t, fetch.PathRE)
assert.Equal(t, ".*", *fetch.PathRE)
```
if you think that reads better.

I use stringPtr and boolPtr in a lot of other tests (deep inside structs, e.g. TestMinimalValidConfig); what would you propose in those cases?

FYI, the reason I'm not just doing a whole struct comparison in this test is that I need to use `ElementsMatch` to ignore sort-order; as the default fetch scheme gets the keys from a map, which are in random order. (I could fix that in prod code by sorting, but I avoid fixing prod for the sake of test when reasonable.)

Reply:
With that explanation, sg to keep what you have.